### ISSUE TITLE
Breaking the Beam Search Curse: A Study of (Re-)Scoring Methods and Stopping Criteria for Neural Machine Translation [Yang et al. 2018]

### ISSUE THREAD

**Issue by zaemyung** [zaemyung/paper-notes#3]
Link: https://arxiv.org/abs/1808.09582
## TL;DR
- Problem
  - Contrary to SMT, the quality of translation degrades with beam sizes beyond 5 or 10, i.e. "beam search curse"
- Proposed Explanation
  1. With increasing beam size, length of generated target sentence (`|y|`) decreases, causing length ratio (`lr`) to drop
  2. This, in turn, severely decreases the brevity penalty term which is a function of the length ratio
  3. Why does `|y|` decrease?
  Increasing beam sizes → the more candidates explored → easier for the search algorithm to find the `</eos>` symbol → shorter candidates tend to have greater model scores, hence selected as top beams
- Approach
  - Three rescoring methods: Bounded Word-Reward, Bounded Adaptive-Reward, BP-Norm
    - All three methods use predicted "generation ratio" learned by utilizing 2-layer MLP which takes the mean of source hidden states as input
  - Optimal Stopping Criteria for the three methods as well as commonly used length norm. method
- Contribution
  - Experimental results show that the rescoring method + stopping criterion improves BLEU even when the beam size is large
### Decreasing BLEU against increasing beam sizes
![screen shot 2018-11-14 at 12 55 37 pm](https://user-images.githubusercontent.com/3746478/48459280-edb34300-e80c-11e8-9c41-71cfd2e76e5d.png)
### BLEU metric:
![screen shot 2018-11-14 at 12 55 23 pm](https://user-images.githubusercontent.com/3746478/48459276-eb50e900-e80c-11e8-9fd1-ef0a93a0f20d.png)
### Previous methods
- Length norm.:
![screen shot 2018-11-14 at 1 43 22 pm](https://user-images.githubusercontent.com/3746478/48461225-0a07ad80-e816-11e8-9d95-6c849ad6190a.png)
- Word reward:
![screen shot 2018-11-14 at 1 43 40 pm](https://user-images.githubusercontent.com/3746478/48461236-1b50ba00-e816-11e8-8a50-9713d1e3a408.png)
- Bounded word reward:
![screen shot 2018-11-14 at 1 43 50 pm](https://user-images.githubusercontent.com/3746478/48461264-34596b00-e816-11e8-81b6-5cf0ac722508.png)
![screen shot 2018-11-14 at 2 00 38 pm](https://user-images.githubusercontent.com/3746478/48461282-4e934900-e816-11e8-82a9-4439009b6575.png)
### Proposed methods
- Length prediction using 2-layer MLP:
![screen shot 2018-11-14 at 2 00 48 pm](https://user-images.githubusercontent.com/3746478/48461321-7b476080-e816-11e8-907f-9352f85e595e.png)
- Bounded word reward with length prediction:
![screen shot 2018-11-14 at 2 00 55 pm](https://user-images.githubusercontent.com/3746478/48461333-8b5f4000-e816-11e8-8903-9d5b9a01de5b.png)
  - Still has `r` as hyp. param.
- Bounded adaptive-reward with length prediction:
![screen shot 2018-11-14 at 2 01 12 pm](https://user-images.githubusercontent.com/3746478/48461382-c5304680-e816-11e8-8782-06bdc7130657.png)
![screen shot 2018-11-14 at 2 01 25 pm](https://user-images.githubusercontent.com/3746478/48461399-d8dbad00-e816-11e8-8a2d-9db27985efad.png)
- BP-norm with length prediction:
![screen shot 2018-11-14 at 2 01 32 pm](https://user-images.githubusercontent.com/3746478/48461406-ed1faa00-e816-11e8-92f3-4f7c2529a354.png)
  - Predicted length is used as the reference length when calculating `bp`
### Experimental Results
![screen shot 2018-11-14 at 2 12 10 pm](https://user-images.githubusercontent.com/3746478/48461494-60c1b700-e817-11e8-9dc4-1a9c338d974d.png)
![screen shot 2018-11-14 at 2 12 25 pm](https://user-images.githubusercontent.com/3746478/48461497-628b7a80-e817-11e8-8668-a500ab1fd993.png)
![screen shot 2018-11-14 at 2 12 41 pm](https://user-images.githubusercontent.com/3746478/48461498-64553e00-e817-11e8-848d-c3368fc8e402.png)
- Authors suggest to use BP-Norm as it performs similarly while being the simplest.
## Notes
- The notion that larger beam sizes should always result in better-quality translations assumes that the training of the NMT model has been conducted in the "best way possible", i.e. the training data covers the ideal distribution that we are (ever going to be) interested in, and the model has fit the distribution well. However, as this is not the case in reality, there may be some test cases that the model wrongly assigns high scores to "wrong" target sequences, and the beam search with larger beam sizes may have found and chosen them.
- The paper does not say from which dataset the 2-layer MLP is trained. Maybe the training set?
- It would be interesting to see where the increase in BLEU points came from. Did the extended beam search manage to recover some of the previously missing tokens that add to the meaning? Or, did it simply favor the sentences with the same semantics but written in a slightly longer form?
- Sadly, at industrial setting, employing larger beam sizes is practically difficult.
